crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,317 @@
+    }
+    ctx->buf_off = 0;
+    ctx->buf_len = outcome;
+  }
+  return bytes_processed;
+}
+
+static int enc_flush(BIO *b, BIO *next, BIO_ENC_CTX *ctx, int do_final) {
+  GUARD_PTR(b);
+  GUARD_PTR(next);
+  GUARD_PTR(ctx);
+  while (ctx->buf_len > 0) {
+    int outcome = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (outcome <= 0) {
+      BIO_copy_next_retry(b);
+      return outcome;
+    }
+    ctx->buf_off += outcome;
+    ctx->buf_len -= outcome;
+  }
+  ctx->buf_off = 0;
+  ctx->buf_len = 0;
+  if (do_final && !ctx->flushed) {
+    ctx->done = 1;
+    ctx->flushed = 1;
+    ctx->ok = EVP_CipherFinal_ex(ctx->cipher, ctx->buf, &ctx->buf_len);
+    return ctx->ok;
+  }
+  return 1;
+}
+
+static int enc_write(BIO *b, const char *in, int inl) {
+  GUARD_PTR(b);
+  GUARD_PTR(in);
+  int ret = 0;
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  if ((ret = enc_flush(b, next, ctx, /*do_final*/ 0)) <= 0) {
+    return ret;
+  }
+
+  assert(ctx->buf_off == ctx->buf_len && ctx->buf_len == 0);
+  int bytes_processed = 0;
+  int remaining = inl;
+  while (remaining > 0) {
+    if (ctx->buf_len == 0) {
+      ctx->buf_off = 0;
+      int to_encrypt = remaining > ENC_BLOCK_SIZE ? ENC_BLOCK_SIZE : remaining;
+      if (!EVP_EncryptUpdate(ctx->cipher, &ctx->buf[0], &ctx->buf_len,
+                             (uint8_t *)in, to_encrypt)) {
+        ret = bytes_processed;
+        goto out;
+      };
+      bytes_processed += ctx->buf_len;
+      remaining -= ctx->buf_len;
+    }
+    int outcome = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (outcome <= 0) {
+      ret = bytes_processed;
+      goto out;
+    }
+    ctx->buf_off += outcome;
+    ctx->buf_len -= outcome;
+  }
+out:
+  if (ret <= 0) {
+    BIO_copy_next_retry(b);
+  }
+  return bytes_processed;
+}
+
+static long enc_ctrl(BIO *b, int cmd, long num, void *ptr) {
+  GUARD_PTR(b);
+  BIO_ENC_CTX *ctx;
+  long ret = 1;
+  BIO *next;
+
+  ctx = BIO_get_data(b);
+  next = BIO_next(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  switch (cmd) {
+    case BIO_CTRL_RESET:
+      ctx->ok = 1;
+      ctx->flushed = 0;
+      if (!EVP_CipherInit_ex(ctx->cipher, NULL, NULL, NULL, NULL,
+                             EVP_CIPHER_CTX_encrypting(ctx->cipher)))
+        return 0;
+      ret = BIO_ctrl(next, cmd, num, ptr);
+      break;
+    case BIO_CTRL_EOF:
+      if (ctx->done) {
+        ret = 1;
+      } else {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_CTRL_WPENDING:
+    case BIO_CTRL_PENDING:
+      // Calculate number of bytes left to process if we have anything buffered,
+      // else consult underlying BIO.
+      ret = ctx->buf_len - ctx->buf_off;
+      if (ret <= 0) {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_CTRL_FLUSH:
+      enc_flush(b, next, ctx, /*do_final*/ 1);
```

Comment:
Does the return value of `enc_flush` needs to be checked here?

Reply:
ooo good find! i have [a test case](https://github.com/aws/aws-lc/pull/1836/files#diff-4cf34798a14703c26416adc6fe0643a8a226c8a0a1f4b9169c5b70741c9af715R261) for this, but it must have failed on flushing the underlying BIO, not when calling `enc_flush`. fixed.